src/clj/harja/palvelin/integraatiot/velho/varusteet.clj

```diff
@@ -0,0 +1,385 @@
+      (do (log/error "Kohteen haku Velhosta epäonnistui. Ei saatu autorisaatio tokenia. Kohteiden url: " url)
+          false))))
+
+(defn muodosta-oidit-url
+  "`kohdeluokka` sisältää /-merkin. esim. `varusteet/kaiteet`"
+  [lahde varuste-api-juuri viimeksi-haettu-velhosta]
+  (let [viimeksi-haettu-iso-8601 (varuste-vastaanottosanoma/aika->velho-aika viimeksi-haettu-velhosta)]
+    (str varuste-api-juuri "/" (:palvelu lahde) "/api/" (:api-versio lahde)
+         "/tunnisteet/" (:kohdeluokka lahde) "?jalkeen="
+         (http/url-encode viimeksi-haettu-iso-8601))))
+
+
+(defn hae-ja-tallenna [lahde viimeksi-haettu-velhosta konteksti varuste-api-juuri token-fn tallenna-fn tallenna-hakuaika-fn]
+  (let [url (muodosta-oidit-url lahde varuste-api-juuri viimeksi-haettu-velhosta)
+        token (token-fn)]
+    (if token
+      (try+
+        (let [haku-alkanut (pvm/nyt)
+              otsikot {"Content-Type" "application/json"
+                       "Authorization" (str "Bearer " token)}
+              http-asetukset {:metodi :GET
+                              :url url
+                              :otsikot otsikot}
+              {body :body headers :headers} (integraatiotapahtuma/laheta konteksti :http http-asetukset)
+              {tila :tila oidit :oidit} (jasenna-varusteiden-oidit url body headers)]
+          (when (and tila (not-empty oidit))
+            (let [oidit-alijoukot (partition
+                                    +kohde-haku-maksimi-koko+
+                                    +kohde-haku-maksimi-koko+
+                                    nil
+                                    oidit)
+                  osajoukon-haku-fn (fn [oidit-alijoukko]
+                                      (hae-kohdetiedot-ja-tallenna-kohde lahde varuste-api-juuri konteksti token-fn
+                                                                         oidit-alijoukko tallenna-fn))
+                  tulokset (map osajoukon-haku-fn oidit-alijoukot)
+                  kaikki-onnistunut (every? true? tulokset)]
+              (when kaikki-onnistunut
+                (tallenna-hakuaika-fn haku-alkanut)))))
+        (catch [:type virheet/+ulkoinen-kasittelyvirhe-koodi+] {:keys [virheet]}
+          (log/error "Haku Velhosta epäonnistui. Virheet: " virheet)))
+      (log/error "Haku Velhosta epäonnistui. Autorisaatio tokenia ei saatu. Kohdeluokan url: " url))))
+
+(defn sijainti-kohteelle [db {:keys [sijainti alkusijainti loppusijainti] :as kohde}]
+  (let [a (or sijainti alkusijainti)
+        b loppusijainti]
+    (assert (some? a) "`sijainti` tai `alkusijainti` on pakollinen")
+    (let [parametrit {:tie (:tie a)
+                      :aosa (:osa a)
+                      :aet (:etaisyys a)
+                      :losa (or (:osa b) (:osa a))
+                      :let (or (:etaisyys b) (:etaisyys a))}]
+      (:sijainti (first (q-toteumat/varustetoteuman-toimenpiteelle-sijainti db parametrit))))))
+
+(defn tuo-uudet-varustetoteumat-velhosta
```

Comment:
Olisiko tämä funktio mahdollista pilkkoa? Aika monta sisäkkäistä try catch lausetta.

Esim funktio: tallenna-toteuma-fn joka luodaan letissä, voisi olla ihan erillinen funktio, jolloin sille voisi tehdä yksikkötestin.

Ja jotenkin hankalasti tuo tallennusfunktio välitetään parin funktion läpi johonkin missä sitä kutsutaan.
En osaa näin äkkikseltään sanoa, että voiko tuon tehdä eri tavalla, mutta näin katselmoidessa koodin lukeminen on hankalaa.

Reply:
Olisiko tämä helpompi lukea?
![image](https://user-images.githubusercontent.com/703699/146939063-6215c4c9-3908-4a2a-adb4-5a1a502f9bd6.png)